File: libUtil/src/main/java/org/waynezhou/libUtil/view_transition/ViewTransition.java
package org.waynezhou.libUtil.view_transition;

import android.animation.Animator;
import android.animation.AnimatorSet;
import android.view.View;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class ViewTransition<TValueHolder> {

    public static class Builder<TValueHolder>{
        private final TValueHolder valHolder;
        protected final List<ViewStep<TValueHolder>> viewSteps = new ArrayList<>();
        public Builder(TValueHolder valHolder) {
            this.valHolder = valHolder;
        }

        public ViewStep.Builder<TValueHolder> startAddStep(ViewGetter viewGetter){
            return new ViewStep.Builder<>(this, viewGetter);
        }
        public ViewTransition<TValueHolder> build(){
            return new ViewTransition<TValueHolder>(valHolder, Collections.unmodifiableList(viewSteps));
        }
    }

    private final TValueHolder valHolder;
    private final List<ViewStep<TValueHolder>> viewSteps;
    private ViewTransition(TValueHolder valHolder, List<ViewStep<TValueHolder>> viewSteps) {
        this.valHolder = valHolder;
        this.viewSteps = viewSteps;
    }

    public AnimatorSet createAnimatorSet(ViewAnimatorArgs args){
        AnimatorSet set = new AnimatorSet();
        List<Animator> animators = new ArrayList<>();
        for(int i =0;i < viewSteps.size();i++){
            animators.add(viewSteps.get(i).createAnimator(valHolder, args));
        }
        set.playTogether(animators);
        return set;
    }

    public void runWithoutAnimation(){
        for(int i =0;i < viewSteps.size();i++){
            viewSteps.get(i).createRunnable(valHolder).run();
        }
    }

    public interface ViewGetter{
        View get();
    }

}

File: libViewKt/src/main/java/org/waynezhou/libviewkt/view_transition/ViewTransition.kt
package org.waynezhou.libviewkt.view_transition

import android.animation.*
import android.view.View
import android.view.animation.AccelerateDecelerateInterpolator
import androidx.core.animation.addListener
import org.waynezhou.libviewkt.view_transition.ViewStep.ViewStepExecutor
import java.util.*

/** Use getter to store layout of views. */
class ViewTransition<TValueHolder : Any> private constructor(
    private val valHolder: TValueHolder,
    private val viewSteps: List<ViewStep<TValueHolder>>,
    private val conditionalViewSteps: List<ConditionalViewStep<TValueHolder>>,
) {
    /** Builder of [ViewTransition]. */
    class Builder<TValueHolder : Any> private constructor(
        private val valHolder: TValueHolder,
        internal val viewSteps: MutableList<ViewStep<TValueHolder>>,
        internal val conditionalViewSteps: MutableList<ConditionalViewStep<TValueHolder>>,
    ) {
        /** Empty Instance Builder Constructor. */
        constructor(valHolder: TValueHolder) : this(valHolder, mutableListOf(), mutableListOf())

        /** Copy Instance Builder Constructor. */
        constructor(other: ViewTransition<TValueHolder>) : this(
            other.valHolder,
            other.viewSteps.toMutableList(),
            other.conditionalViewSteps.toMutableList()
        )

        /** Add a transition step of a view. */
        fun <TView:View> beginAddStep(getter: () -> TView): ViewStep.Builder<TValueHolder, TView> {
            return ViewStep.Builder(this, getter)
        }

        /** Add a transition step of a view witch will be triggered in conditional. */
        fun <TView:View> beginAddStep(condition: () -> Boolean, getter: () -> TView) : ViewStep.Builder<TValueHolder, TView> {
            return ViewStep.Builder(this, condition, getter)
        }

        /** Build new [ViewTransition] instance. */
        fun build(): ViewTransition<TValueHolder> {
            return ViewTransition(valHolder, viewSteps.toList(), conditionalViewSteps.toList())
        }
    }

    /** Return new [AnimatorSet] instance that animate layout with captured value from getter. */
    fun createAnimatorSet(args: ViewAnimatorArgs): AnimatorSet {
        val set = AnimatorSet()
        val animators = mutableListOf<Animator>()

        viewSteps.forEach { animators.add(it.createAnimator(valHolder, args)) }
        conditionalViewSteps.filter { it.condition() }
            .forEach { animators.add(it.viewStep.createAnimator(valHolder, args)) }
        set.playTogether(animators)
        return set
    }

    /** Change layout immediately without animation. */
    fun runWithoutAnimation(withRequestLayout: Boolean) {
        viewSteps.forEach {
            it.createExecutor(valHolder).execute(withRequestLayout)
        }
        conditionalViewSteps.filter { it.condition() }
            .forEach {
                it.viewStep.createExecutor(valHolder).execute(withRequestLayout)
            }
    }

    fun transpose(): ViewTransition<TValueHolder>{
        return ViewTransition(
            valHolder,
            viewSteps.map{it.transpose()},
            conditionalViewSteps.map { ConditionalViewStep(it.condition, it.viewStep.transpose()) }
        )
    }


}

data class ViewAnimatorArgs constructor(
    val duration: Long = 500,
    val interpolator: TimeInterpolator = AccelerateDecelerateInterpolator(),
)
